From: Sean Young <sean@mess.org>
Date: Mon, 13 Aug 2012 08:59:40 -0300
Subject: [03/11] [media] iguanair: ignore unsupported firmware versions

commit 0797b4802b57ce88e9361c8d64b0980508370649 upstream.

Firmware versions lower than 0x0205 use a different interface which is not
supported. Also report the firmware version in the standard format.

Signed-off-by: Sean Young <sean@mess.org>
Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>
---
 drivers/media/rc/Kconfig    |    8 ++++++--
 drivers/media/rc/iguanair.c |   21 +++++++++++----------
 2 files changed, 17 insertions(+), 12 deletions(-)

Index: linux-3.2.46/drivers/media/rc/Kconfig
===================================================================
--- linux-3.2.46.orig/drivers/media/rc/Kconfig	2013-07-26 19:25:50.000000000 +0000
+++ linux-3.2.46/drivers/media/rc/Kconfig	2013-07-26 19:25:52.000000000 +0000
@@ -248,8 +248,12 @@
 	depends on RC_CORE
 	select USB
 	---help---
-	   Say Y here if you want to use the IgaunaWorks USB IR Transceiver.
-	   Both infrared receive and send are supported.
+	   Say Y here if you want to use the IguanaWorks USB IR Transceiver.
+	   Both infrared receive and send are supported. If you want to
+	   change the ID or the pin config, use the user space driver from
+	   IguanaWorks.
+
+	   Only firmware 0x0205 and later is supported.
 
 	   To compile this driver as a module, choose M here: the module will
 	   be called iguanair.
Index: linux-3.2.46/drivers/media/rc/iguanair.c
===================================================================
--- linux-3.2.46.orig/drivers/media/rc/iguanair.c	2013-07-26 19:25:51.000000000 +0000
+++ linux-3.2.46/drivers/media/rc/iguanair.c	2013-07-26 19:25:53.000000000 +0000
@@ -36,8 +36,8 @@
 	struct usb_device *udev;
 
 	int pipe_out;
+	uint16_t version;
 	uint8_t bufsize;
-	uint8_t version[2];
 
 	struct mutex lock;
 
@@ -97,8 +97,8 @@
 		switch (ir->buf_in[3]) {
 		case CMD_GET_VERSION:
 			if (len == 6) {
-				ir->version[0] = ir->buf_in[4];
-				ir->version[1] = ir->buf_in[5];
+				ir->version = (ir->buf_in[5] << 8) |
+							ir->buf_in[4];
 				complete(&ir->completion);
 			}
 			break;
@@ -110,8 +110,7 @@
 			break;
 		case CMD_GET_FEATURES:
 			if (len > 5) {
-				if (ir->version[0] >= 4)
-					ir->cycle_overhead = ir->buf_in[5];
+				ir->cycle_overhead = ir->buf_in[5];
 				complete(&ir->completion);
 			}
 			break;
@@ -219,6 +218,12 @@
 		goto out;
 	}
 
+	if (ir->version < 0x205) {
+		dev_err(ir->dev, "firmware 0x%04x is too old\n", ir->version);
+		rc = -ENODEV;
+		goto out;
+	}
+
 	ir->bufsize = 150;
 	ir->cycle_overhead = 65;
 
@@ -230,9 +235,6 @@
 		goto out;
 	}
 
-	if (ir->version[0] == 0 || ir->version[1] == 0)
-		goto out;
-
 	packet.cmd = CMD_GET_FEATURES;
 
 	rc = iguanair_send(ir, &packet, sizeof(packet));
@@ -485,8 +487,7 @@
 		goto out2;
 
 	snprintf(ir->name, sizeof(ir->name),
-		"IguanaWorks USB IR Transceiver version %d.%d",
-		ir->version[0], ir->version[1]);
+		"IguanaWorks USB IR Transceiver version 0x%04x", ir->version);
 
 	usb_make_path(ir->udev, ir->phys, sizeof(ir->phys));
 
